packages/muelu/research/regionMG/src/SetupRegionMatrix_def.hpp

```diff
@@ -167,41 +162,29 @@ void MakeQuasiregionMatrices(const RCP<Xpetra::CrsMatrixWrap<Scalar, LocalOrdina
                             *(rowImport),
                             Xpetra::INSERT);
 
-  quasiRegionMats->fillComplete();
-
   tm = Teuchos::null;
   tm = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("MakeQuasiregionMatrices: 3 - Scale interface entries")));
 
   RCP<CrsMatrixWrap> quasiRegionCrsWrap = Teuchos::rcp_dynamic_cast<CrsMatrixWrap>(quasiRegionMats);
   RCP<CrsMatrix> quasiRegionCrs = quasiRegionCrsWrap->getCrsMatrix();
-  const LO numRows    = Teuchos::as<LO>(quasiRegionCrs->getNodeNumRows());
-  // const size_t numNNZ = quasiRegionCrs->getNodeNumEntries();
-
-  ArrayRCP<const size_t> rowPtrCopy;
-  ArrayRCP<const LO> colIndsCopy;
-  ArrayRCP<const SC> valuesCopy;
-  quasiRegionCrs->getAllValues(rowPtrCopy, colIndsCopy, valuesCopy);
-
-  *myOut << "Just extracted quasi region matrix data, numEntries="
-         << valuesCopy.size() << ", last entry ID: " << rowPtrCopy[numRows]
-         << ", colIndsCopy.size()=" << colIndsCopy.size()
-         << ", valuesCopy.size()=" << valuesCopy.size() << std::endl;
-
-  ArrayRCP<size_t> rowPtr(rowPtrCopy.size());
-  ArrayRCP<LO> colInds(colIndsCopy.size());
-  ArrayRCP<SC> values(valuesCopy.size());
-  rowPtr.deepCopy(rowPtrCopy.view(0, rowPtrCopy.size()));
-  colInds.deepCopy(colIndsCopy.view(0, colIndsCopy.size()));
-  values.deepCopy(valuesCopy.view(0, valuesCopy.size()));
+
+  Array<LO> tmp(regionMatVecLIDs());
+  std::sort(tmp.begin(), tmp.end());
+  auto vecEnd = std::unique(tmp.begin(), tmp.end());
+  auto vecStart = tmp.begin();
 
   GO rowGID;
   LocalOrdinal col;
   GlobalOrdinal colGID;
   std::size_t sizeOfCommonRegions;
-  for (LO row = 0; row < numRows; ++row) { // loop over local rows of composite matrix
-    rowGID = rowMap->getGlobalElement(row);
-
-    for (std::size_t entryIdx = rowPtr[row]; entryIdx < rowPtr[row + 1]; ++entryIdx) { // loop over all entries in this row
+  for(auto row = vecStart; row < vecEnd; ++row) {
+    rowGID = rowMap->getGlobalElement(*row);
+    std::size_t numEntries = quasiRegionMats->getNumEntriesInLocalRow(*row); // number of entries in this row
```

Comment:
Move this declaration out of the `for`-loop to only allocate memory once? There are already a bunch of declarations just prior to the `for`-loop.